### PR TITLE
DEV-5523: add authed payments endpoint for switchboard

### PR DESCRIPTION
**Pull request by leo-dcfa**
Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

N/A

#### What's this PR do?

- Added an authenticated payments endpoint for Switchboard.
- Made `transaction_time` on `Payment` model required
- Removed `sync_payment_transaction_time.py` as it is no longer needed

#### Why are we doing this? How does it help us?

- Part of general efforts to allow Switchboard to create entities in revengine based on events driven by other systems.
- `transaction_time` - we should never allow a payment to have a null-value for transaction time. We have ran a management command to make sure transaction_time is always populated and it should be required going forward
- `sync_payment_transaction_time` is no longer in use, since  `transaction_time` is now required

#### Are there detailed, specific, step-by-step testing instructions in the associated ticket?

Yes.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No

#### Have automated unit tests been added? If not, why?

Yes

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

No

#### Has this been documented? If so, where?

N/A

#### What are the relevant tickets? Add a link to any relevant ones.

[DEV-5523](https://news-revenue-hub.atlassian.net/browse/DEV-5523)

#### Do any changes need to be made before deployment to production (adding environment variables, for example)? If so, open a ticket and link it to the ticket for this PR and list it here:

Migration on `transaction_time` might take a while.

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

No

[DEV-5523]: https://news-revenue-hub.atlassian.net/browse/DEV-5523?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/contributions/serializers.py [hunk 1]
```diff
@@ -1026,3 +1026,18 @@ class Meta:
...
+
+class SwitchboardPaymentSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Payment
+        fields = [
+            "id",
+            "contribution",
+            "net_amount_paid",
+            "gross_amount_paid",
+            "amount_refunded",
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/contributions/serializers.py:1039]: Adding to this: we'll want to do this additional validation by overriding the serializer `.validate` method.  That runs after field level validation has been happened and is the appropriate place for doing validation that involves looking at > 1 field in relation to one another.

You can see an example of this in `PortalContributionDetailSerializer.validate` and elsewhere in revengine.

[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/contributions/serializers.py:1039]: While we're at it, we probably should consider requiring that if gross amount is > 0 so to is net amount.

File: apps/contributions/views/switchboard.py [hunk 2]
```diff
@@ -91,3 +91,32 @@ def contributor_by_email(request: HttpRequest, email: str) -> Response:
     contributor = get_object_or_404(Contributor.objects.all(), email=email)
     serializer = serializers.SwitchboardContributorSerializer(contributor)
     return Response(serializer.data)
+
+
+class SwitchboardPaymentsViewSet(
```
Comment (benjaminEwhite) [inline apps/contributions/views/switchboard.py:96]: In case that a create/update operation would cause an uniqueness constraint integrity error on stripe_balance_transaction_id, I think we should return a 409 so that Switchboard can easily infer that it needs to update an extant payment.

I think we can move the custom overridde that I recently added to `SwitchboardContributionsViewSet.handle_exception`  and pull that into a mixin, that both `SwitchboardContributionsViewSet` and `SwitchboardPaymentsViewSet` can inherit from.  Pasting in that code here for context:

```
def handle_exception(self, exc):
        """Ensure select uniqueness constraint errors receive a 409.
        """
        if isinstance(exc, ValidationError):
            details = exc.detail
            for errors in details.values():
                if any(x.code == "unique" for x in errors):
                    exc.status_code = status.HTTP_409_CONFLICT
                    break
        return super().handle_exception(exc)
```

File: apps/contributions/tests/views/test_switchboard.py [hunk 3]
```diff
@@ -537,3 +539,128 @@ def test_requests_when_token_expired(self, api_client, expired_token, method):
...
         assert response.status_code == status.HTTP_401_UNAUTHORIZED
         assert response.json() == {"detail": "Invalid token."}
+
+
+@pytest.mark.django_db
+class TestSwitchboardPaymentsViewSet:
+
+    @pytest.fixture
+    def super_user(self):
+        return UserFactory(is_superuser=True)
```
Comment (benjaminEwhite) [inline apps/contributions/tests/views/test_switchboard.py:549]: Think we can reuse existing `superuser` fixture from top level conftest.py.

File: apps/contributions/tests/views/test_switchboard.py [hunk 4]
```diff
@@ -537,3 +539,128 @@ def test_requests_when_token_expired(self, api_client, expired_token, method):
...
+
+    @pytest.fixture
+    def contribution(self):
+        return ContributionFactory()
+
+    @pytest.fixture
+    def payment(self, contribution):
+        return PaymentFactory(contribution=contribution)
```
Comment (benjaminEwhite) [inline apps/contributions/tests/views/test_switchboard.py:557]: `PaymentFactory` already creates a contribution by default so this can either be simplified, or could just use PaymentFactory inline in test without fixture.

File: apps/contributions/tests/views/test_switchboard.py [hunk 5]
```diff
@@ -537,3 +539,128 @@ def test_requests_when_token_expired(self, api_client, expired_token, method):
...
+            headers={"Authorization": f"Token {token}"},
+            format="json",
+        )
+        assert response.status_code == status.HTTP_200_OK
+        assert response.data["contribution"] == contribution.id
+        assert response.data["net_amount_paid"] == 2500
+        assert response.data["gross_amount_paid"] == 2700
+        assert response.data["amount_refunded"] == 500
+        assert response.data["stripe_balance_transaction_id"] == "txn_987654"
+        assert response.data["transaction_time"] == "2025-02-01T22:30:45.654321Z"
```
Comment (benjaminEwhite) [inline apps/contributions/tests/views/test_switchboard.py:666]: This can be removed once we prohibit PUT.

File: apps/contributions/tests/views/test_switchboard.py [hunk 6]
```diff
@@ -537,3 +539,128 @@ def test_requests_when_token_expired(self, api_client, expired_token, method):
...
+        url_kwargs = {"args": (payment.id,)} if method == "get" else {}
+        url = reverse(url_name, **url_kwargs)
+        request_kwargs = {"data": {}} if method == "post" else {}
+        response = getattr(api_client, method)(
+            url, headers={"Authorization": f"Token {expired_token}", **request_kwargs}
+        )
+        assert response.status_code == status.HTTP_401_UNAUTHORIZED
+        assert response.json() == {"detail": "Invalid token."}
+
+    def test_patch_payment(self, api_client, token, payment):
```
Comment (benjaminEwhite) [inline apps/contributions/tests/views/test_switchboard.py:631]: Would like to see an additional test that demonstrates behavior if you update an existing payment such that it would violate uniqueness constraint on stripe_balance_transaction_id.

File: apps/contributions/tests/test_serializers.py [hunk 7]
```diff
@@ -1790,3 +1791,60 @@ def test_get_revenue_program_source_when_no_source(self, one_time_contribution):
...
+
+    def test_serializer_valid_data(self, contribution):
+        data = {
+            "contribution": contribution.id,
+            "net_amount_paid": 2000,
+            "gross_amount_paid": 2000,
+            "amount_refunded": 0,
+            "stripe_balance_transaction_id": "txn_123456",
+            "transaction_time": timezone.now().isoformat(),
+        }
```
Comment (benjaminEwhite) [inline apps/contributions/tests/test_serializers.py:1815]: Should be able to do this without need for contribution fixture. Just use the payment fixture, pass it to the test, and can use payment.contribution.

**Review by benjaminEwhite — CHANGES_REQUESTED** [review]

File: apps/api/urls.py [hunk 8]
```diff
@@ -27,6 +28,7 @@
 switchboard_router = DefaultRouter()
 switchboard_router.register(r"contributions", SwitchboardContributionsViewSet, basename="switchboard-contribution")
 switchboard_router.register(r"contributors", SwitchboardContributorsViewSet, basename="switchboard-contributor")
+switchboard_router.register(r"payments", SwitchboardPaymentsViewSet, basename="switchboard-payment")
```
[the remark this answers is not in the document]
Comment (leo-dcfa) [reply apps/api/urls.py:31]: @benjaminEwhite I'm a bit confused about this request, the code in your comment matches exactly what's in the code I added which is to add a payments url to switchboard_router. Do you mind clarifying?

File: apps/contributions/tests/views/test_switchboard.py (hunk 5 above)
Comment (leo-dcfa) [reply apps/contributions/tests/views/test_switchboard.py]: Removed https://github.com/newsrevenuehub/rev-engine/pull/1716/commits/b36cce3b7554ad3726a7ea456b261cb3a4757e31

File: apps/contributions/views/switchboard.py [hunk 9]
```diff
@@ -91,3 +91,32 @@ def contributor_by_email(request: HttpRequest, email: str) -> Response:
...
+    mixins.RetrieveModelMixin, mixins.CreateModelMixin, mixins.UpdateModelMixin, viewsets.GenericViewSet
+):
+    """ViewSet for switchboard to create and retrieve payments."""
+
+    permission_classes = [IsSwitchboardAccount]
+    http_method_names = ["get", "post", "patch", "put"]
```
[the remark this answers is not in the document]
Comment (leo-dcfa) [reply apps/contributions/views/switchboard.py]: Removed https://github.com/newsrevenuehub/rev-engine/pull/1716/commits/b36cce3b7554ad3726a7ea456b261cb3a4757e31

File: apps/contributions/views/switchboard.py [hunk 10]
```diff
@@ -91,3 +91,32 @@ def contributor_by_email(request: HttpRequest, email: str) -> Response:
     contributor = get_object_or_404(Contributor.objects.all(), email=email)
     serializer = serializers.SwitchboardContributorSerializer(contributor)
     return Response(serializer.data)
+
+
+class SwitchboardPaymentsViewSet(
+    mixins.RetrieveModelMixin, mixins.CreateModelMixin, mixins.UpdateModelMixin, viewsets.GenericViewSet
+):
+    """ViewSet for switchboard to create and retrieve payments."""
```
[the remark this answers is not in the document]
Comment (leo-dcfa) [reply apps/contributions/views/switchboard.py]: Good catch https://github.com/newsrevenuehub/rev-engine/pull/1716/commits/79ef16a6d3a0f8b1a9ddcde3302155978c27e9eb

File: apps/contributions/serializers.py [hunk 11]
```diff
@@ -1026,3 +1026,18 @@ class Meta:
...
+            "stripe_balance_transaction_id",
+            "transaction_time",
```
[the remark this answers is not in the document]
Comment (leo-dcfa) [reply apps/contributions/serializers.py:1041]: I think changing the model is the right call. If we already intend to change the model, adding the extra validation logic to the serialiser is creating more work for us later.

https://github.com/newsrevenuehub/rev-engine/pull/1716/commits/4b29d1afd97eedc8b8029493f9afe26b5c84bda5

With the knowledge that we have 0 records in prod with a null incorrect transaction_time, I made the default `now` but do you have a different preference?

File: apps/contributions/tests/test_serializers.py (hunk 7 above)
Comment (leo-dcfa) [reply apps/contributions/tests/test_serializers.py]: Good point https://github.com/newsrevenuehub/rev-engine/pull/1716/commits/381e4f5485defc6a3b7763637cb1bfe606d0f36b

File: apps/contributions/tests/views/test_switchboard.py (hunk 4 above)
Comment (leo-dcfa) [reply apps/contributions/tests/views/test_switchboard.py:553]: Thanks! https://github.com/newsrevenuehub/rev-engine/pull/1716/commits/0c72aa3b3f680968e8c1e58e8210dca9ad81df07

File: apps/contributions/views/switchboard.py [hunk 12]
```diff
@@ -91,3 +91,32 @@ def contributor_by_email(request: HttpRequest, email: str) -> Response:
...
+            )
+        ).exists():
+
+            return Response(
+                {
+                    "error": f"A payment (ID: {(_first:=existing.first()).id}) "
+                    f"with stripe_balance_transaction_id {_first.stripe_balance_transaction_id} already exists"
+                },
+                status=status.HTTP_409_CONFLICT,
+            )
```
[the remark this answers is not in the document]
Comment (leo-dcfa) [reply apps/contributions/views/switchboard.py]: I prefer that too, thanks for suggesting it. https://github.com/newsrevenuehub/rev-engine/pull/1716/commits/851d7a50a0ad98017c527c002735daf8d97128ac
I was thrown off by `SwitchboardContributorsViewSet` where we are overriding the `.create` method, stemming from my limited Django knowledge, I went looking for similar examples and this one was in the same file.

File: apps/contributions/tests/views/test_switchboard.py (hunk 6 above)
Comment (leo-dcfa) [reply apps/contributions/tests/views/test_switchboard.py]: https://github.com/newsrevenuehub/rev-engine/pull/1716/commits/828422c729378f88c530deaa21af818adde90b51

File: apps/contributions/tests/views/test_switchboard.py (hunk 3 above)
Comment (leo-dcfa) [reply apps/contributions/tests/views/test_switchboard.py]: Thanks, using `superuser` now https://github.com/newsrevenuehub/rev-engine/pull/1716/commits/0c72aa3b3f680968e8c1e58e8210dca9ad81df07

File: apps/contributions/views/switchboard.py (hunk 2 above)
Comment (leo-dcfa) [reply apps/contributions/views/switchboard.py:82]: great idea https://github.com/newsrevenuehub/rev-engine/pull/1716/commits/6e1d56689c260bad346cf063c7e103596537d506

File: apps/contributions/serializers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (leo-dcfa) [reply apps/contributions/serializers.py:1039]: Thanks for pointing out the usage of `validate` method, I change the initial validation of stripe_balance_transaction_id to its own method + added new methods for net amount and gross amount as well https://github.com/newsrevenuehub/rev-engine/pull/1716/commits/b8f150c99cef38fb53e22fc566566f67b231d8aa

File: apps/contributions/serializers.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (leo-dcfa) [reply apps/contributions/serializers.py:1041]: Actually I realise having a default value kinda defeats the purpose of making the field required.... I revert it back here https://github.com/newsrevenuehub/rev-engine/pull/1716/commits/2bc4ab5cac064238ea95bf4317c067be43ecfaef but keen to hear what you think.

[the remark this answers is not in the document]
Comment (leo-dcfa) [reply apps/contributions/serializers.py:1041]: Also to note, there are a few tests under class `Test_sync_payment_transaction_time` that all set transaction_time to None. Thoughts on what we should do with those tests?

[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/contributions/serializers.py:1041]: Good point! I think we can just get rid of that management command and its tests. It only touches payments that don't have transaction time, and by definition moving forward, all will.

File: apps/contributions/signals.py [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+from apps.contributions.models import Payment
+
+
+logger = logging.getLogger(f"{settings.DEFAULT_LOGGER}.{__name__}")
+
+
+@receiver(post_save, sender=Payment)
+def handle_payment_post_save(sender, instance: Payment, created: bool, **kwargs) -> None:
+    logger.debug("handle_payment_post_save called on payment %s", instance.id)
+    instance.publish_payment_created() if created else instance.publish_payment_updated()
```
Comment (benjaminEwhite) [inline apps/contributions/signals.py:16]: Signals get a bad wrap in Django world (personally I haven't encountered problems with them, but they have a less than stellar reputation).  This further convinces me that we should hold off on emitting events in this PR and circle back to it in a separate one.

File: apps/contributions/tests/test_serializers.py [hunk 14]
```diff
@@ -1790,3 +1791,94 @@ def test_get_revenue_program_source_when_no_source(self, one_time_contribution):
...
+            "net_amount_paid",
+            "gross_amount_paid",
+            "amount_refunded",
+            "stripe_balance_transaction_id",
+            "transaction_time",
+        }
+
+    def test_serializer_missing_required_fields(self):
+        data = {}
+        serializer = SwitchboardPaymentSerializer(data=data)
```
Comment (benjaminEwhite) [inline apps/contributions/tests/test_serializers.py:1828]: Minor but no need to assign data to a var here, can just pass an empty dict directly.

File: apps/contributions/tests/test_serializers.py [hunk 15]
```diff
@@ -1790,3 +1791,94 @@ def test_get_revenue_program_source_when_no_source(self, one_time_contribution):
...
+        assert not serializer.is_valid()
```
Comment (benjaminEwhite) [inline apps/contributions/tests/test_serializers.py:1829]:
```suggestion
        assert serializer.is_valid() is False
```

File: apps/contributions/tests/test_serializers.py [hunk 16]
```diff
@@ -1790,3 +1791,94 @@ def test_get_revenue_program_source_when_no_source(self, one_time_contribution):
...
+        data = {
+            "contribution": 99999,
+            "net_amount_paid": 2000,
+            "gross_amount_paid": 2000,
+            "amount_refunded": 0,
+            "stripe_balance_transaction_id": "txn_123456",
+            "transaction_time": timezone.now().isoformat(),
+        }
+        serializer = SwitchboardPaymentSerializer(data=data)
+        assert not serializer.is_valid()
```
Comment (benjaminEwhite) [inline apps/contributions/tests/test_serializers.py:1849]:
```suggestion
        assert serializer.is_valid() is False
```

File: apps/contributions/tests/test_serializers.py [hunk 17]
```diff
@@ -1790,3 +1791,94 @@ def test_get_revenue_program_source_when_no_source(self, one_time_contribution):
...
+        data = {
+            "contribution": payment.contribution.id,
+            "net_amount_paid": 2000,
+            "gross_amount_paid": 2000,
+            "amount_refunded": 0,
+            "stripe_balance_transaction_id": payment.stripe_balance_transaction_id,
+            "transaction_time": timezone.now().isoformat(),
+        }
+        serializer = SwitchboardPaymentSerializer(data=data)
+        assert not serializer.is_valid()
```
Comment (benjaminEwhite) [inline apps/contributions/tests/test_serializers.py:1862]: Ditto

**Review by benjaminEwhite — CHANGES_REQUESTED** [review]

File: apps/api/mixins.py [hunk 18]
```diff
@@ -0,0 +1,19 @@
+from rest_framework import status
+from rest_framework.exceptions import ValidationError
+
+
+class UniquenessConstraintViolationMixin:
```
[the remark this answers is not in the document]
Comment (leo-dcfa) [reply apps/api/mixins.py]: Good idea 4d6a7f2397d1e5ee99bf9c9c93fd7de4c7238fad

File: apps/contributions/models.py [hunk 19]
```diff
@@ -1405,7 +1411,7 @@ class Payment(IndexedTimeStampedModel):
     # Stripe payment intent, invoice, or charge associated with the balance transaction that has a .created property. We look
     # to balance transaction since it is common to each of: one-time payment, recurring payment, and refund.
     # Ultimately, this field gives us a way to sort by recency.
-    transaction_time = models.DateTimeField(db_index=True, null=True)
+    transaction_time = models.DateTimeField(db_index=True, blank=False, null=False)
```
[the remark this answers is not in the document]
Comment (leo-dcfa) [reply apps/contributions/models.py:1413]: Good catch https://github.com/newsrevenuehub/rev-engine/pull/1716/commits/c4d837a1803edc37edeccc4dfd2a18bbb091b518

File: apps/contributions/signals.py (hunk 13 above)
Comment (leo-dcfa) [reply apps/contributions/signals.py]: one of the things that surprised me about Django Signals is their lack of reliability/lack of retry support. I think something a bit more reliable like we have in SB would be ideal, where we get alerted when things go wrong
Reverted commit here https://github.com/newsrevenuehub/rev-engine/pull/1716/commits/f2ec70a61b1144b7508bae2698118fbd3d47383c

File: apps/contributions/tests/test_serializers.py (hunk 14 above)
Comment (leo-dcfa) [reply apps/contributions/tests/test_serializers.py]: https://github.com/newsrevenuehub/rev-engine/pull/1716/commits/70fcc1801b04a96679d1cfdee6adfa9cafad8862

File: apps/contributions/tests/test_serializers.py (hunk 17 above)
Comment (leo-dcfa) [reply apps/contributions/tests/test_serializers.py]: bc9dedba47df50875e450c33b6f26fe907e2fd92 fixed it here along with other assertions in this file

File: apps/contributions/tests/views/test_switchboard.py (hunk 4 above)
Comment (leo-dcfa) [reply apps/contributions/tests/views/test_switchboard.py]: Duh, I swear I had done it https://github.com/newsrevenuehub/rev-engine/pull/1716/commits/0fa4105562029f3fb51637989927ceee54fa0b96 going crazy

File: apps/contributions/models.py [hunk 20]
```diff
@@ -1636,3 +1642,21 @@ def from_stripe_invoice_payment_succeeded_event(cls, event: StripeEventData) ->
...
+            Publisher.get_instance().publish(settings.CREATE_PAYMENT_TOPIC, Message(data=str(self.id)))
+
+    def publish_payment_updated(self):
+        logger.info(
+            "Publishing UPDATE_PAYMENT_TOPIC (%s) for payment=[%s]",
+            settings.UPDATE_PAYMENT_TOPIC,
+            self.id,
+        )
+        if google_cloud_pub_sub_is_configured():
+            Publisher.get_instance().publish(settings.UPDATE_PAYMENT_TOPIC, Message(data=str(self.id)))
```
[the remark this answers is not in the document]
Comment (leo-dcfa) [reply apps/contributions/models.py]: That's a fair point. I did my best to stick to existing patterns but as noted in the ticket, I don't think it's the right approach. As we grow in more use cases, a more established pattern should be used.

File: apps/contributions/serializers.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (leo-dcfa) [reply apps/contributions/serializers.py:1041]: Sounds like we can get rid of it! https://github.com/newsrevenuehub/rev-engine/pull/1716/commits/53cc294230f276a73713aaa72eeb83f52fe3a178

**Review by benjaminEwhite — APPROVED** [review]
One small request to update PR description re: removing sync_payment_transaction_time.py, and another to add a test for an untested path in serializer.

I'm going to approve with understanding that you'll take care of those two issues, but no need for re-review from me (otherwise we lose another 24 cycle to me reviewing one more time).

File: apps/contributions/serializers.py [hunk 21]
```diff
@@ -1026,3 +1026,50 @@ class Meta:
...
+            raise serializers.ValidationError(
+                "Amount refunded cannot be positive when net_amount_paid or gross_amount_paid are positive"
+            )
+        return data
+
+    def validate_stripe_balance_transaction_id(self, value):
+        if existing := Payment.objects.filter(stripe_balance_transaction_id__iexact=value.strip()).exists():
+            raise serializers.ValidationError(
+                f"A payment (ID: {(_first := existing.first()).id}) "
+                f"with stripe_balance_transaction_id {_first.stripe_balance_transaction_id} already exists"
```
[the remark this answers is not in the document]
Comment (leo-dcfa) [reply apps/contributions/serializers.py:1060]: Good catch 58c6eee162e7b1071551339974c23c1542472a71

File: apps/api/mixins.py [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+    def handle_exception(self, exc):
+        """Ensure select uniqueness constraint errors receive a 409.
+
+        For uniqueness constraints we want to return a 409 Conflict status code.
+        On creation in particular, this will signal that a caller needs
+        to update an existing record rather than create a new one.
+        """
+        if isinstance(exc, ValidationError):
+            details = exc.detail
+            for errors in details.values():
```
Comment (njharman) [inline apps/api/mixins.py:15]: exc.detail can be a list (and will be unless value passed into constructor is a dict()). This code should handle list gracefully. Probably by ignoring

```
if isinstance(exc.detail, list):
  continue
```


See https://www.django-rest-framework.org/api-guide/exceptions/#validationerror and source code rest_framework/exceptions.py

File: apps/contributions/serializers.py [hunk 23]
```diff
@@ -1026,3 +1026,50 @@ class Meta:
...
+
+        if amount_refunded > 0 and (net_amount_paid > 0 or gross_amount_paid > 0):
+            raise serializers.ValidationError(
+                "Amount refunded cannot be positive when net_amount_paid or gross_amount_paid are positive"
+            )
+        return data
+
+    def validate_stripe_balance_transaction_id(self, value):
+        if existing := Payment.objects.filter(stripe_balance_transaction_id__iexact=value.strip()).exists():
+            raise serializers.ValidationError(
```
Comment (njharman) [inline apps/contributions/serializers.py:1058]: I don't believe this produces 409 per AC4. https://www.django-rest-framework.org/api-guide/exceptions/#validationerror indicates it produces 400

**Review by njharman — CHANGES_REQUESTED** [review]

File: apps/api/mixins.py (hunk 22 above)
Comment (leo-dcfa) [reply apps/api/mixins.py:15]: https://github.com/newsrevenuehub/rev-engine/pull/1716/commits/1230e312aeefb3b72bdb3329749b4cf87b0138c8

**Review by njharman — APPROVED** [review]

File: apps/api/mixins.py (hunk 22 above)
Comment (njharman) [reply apps/api/mixins.py:15]: Uncovered now

**Review by njharman — CHANGES_REQUESTED** [review]

**Review by njharman — APPROVED** [review]

**Review by x110dc — APPROVED** [review]

File: apps/contributions/views/switchboard.py [hunk 24]
```diff
@@ -91,3 +79,30 @@ def contributor_by_email(request: HttpRequest, email: str) -> Response:
...
+
+    def perform_create(self, serializer):
+        with reversion.create_revision():
+            serializer.save()
+            reversion.set_comment(self.request.data.get("comment", ""))
+
+    def perform_update(self, serializer):
+        with reversion.create_revision():
+            serializer.save()
+            reversion.set_comment(self.request.data.get("comment", ""))
```
Comment (benjaminEwhite) [inline apps/contributions/views/switchboard.py:108]: I'd prefer to have this follow patterns set elsewhere were revision comment is hard coded. So for perform create, we'd do something like "Created by Switchboard API" and for update "Updated by Switchboard API".

We should be able to look for predictable revision comments in order to understand the history of a given object.
